### ISSUE TITLE
Communicate Futures between Clients

### ISSUE THREAD

**Issue by mrocklin** [dask/distributed#701]
How do we communicate futures between clients?

The recent ability of [tasks to spawn other tasks](http://distributed.readthedocs.io/en/latest/task-launch.html) with `local_client` opens the possibility for highly dynamic computations.  For example we could have workers watch live data sources and create feeds of data to which other clients can respond.  This is all possible today except that there is no mechanism for clients to share futures between each other.  They all use the same computation memory resources but have no ability to make each other aware of what is on the cluster.  Several users of `local_client` have noted this limitation.

So what is the right abstraction to communicate futures between clients?

## Option One: Kafka Topics

We could hold Future objects in a rolling buffer.  This would be a data structure living on the centralized scheduler to which all clients would get a view.  Appends would always happen to the end of this buffer while every client would also have a read-head that they could use for iteration.  It's worth noting that, unlike Kafka this wouldn't hold actual data, just `Futures`.  The actual data would live on the workers.

## Option Two: Custom

I've spoken to a few people that have custom needs here, so providing one abstraction like Kafka's topic probably doesn't suffice.  Instead we would want to create an interface that others could implement and upload to the scheduler.  A couple examples that have come up in the wild:

*  A collection into which we would place futures sequentially but would only allow collecting that future fixed number of times
*  A fixed length collection that we would mutate with new futures as new data came in.  Clients watching this collection might re-run a computation on every change.
*  Our current [`get/publish_dataset`](http://distributed.readthedocs.io/en/latest/publish.html) functionality might fall into this

## API Play

So here is a possible API for critique:

### Data Collection

```python
with local_client() as c:
    topic = c.topic('raw-data')
    for batch in watch_data_feed(...):
        future = c.scatter(batch)
        topic.append(future)
```

### Process Data

We pull the raw data futures into two different worker/clients, process with a couple of functions, and then push out to a feed of processed data.

```python
with local_client() as c:
    raw = c.topic('raw-data')
    processed = c.topic('processed-data')
    for future in raw:
        future2 = c.submit(f, future)
        processed.append(future2)
```

```python
with local_client() as c:
    raw = c.topic('raw-data')
    processed = c.topic('processed-data')
    for future in raw:
        future2 = c.submit(g, future)
        processed.append(future2)
```    

### Collect

We get futures from the processed data topic, gather the data to the local machine, and then do something with it here such as emit some plots or analysis.  We pull the results of both processing clients without caring which future was submitted by which client.

```python
with local_client() as c:
    processed = c.topic('processed-data')
    for future in processed:
        data = c.gather(future)
        ...
```

**Comment by strets123** [comment]
I can see that there is real flexibility and either of the above solutions could be made to work for me.

This may be a silly question but I was wondering whether it would be possible to gain future control of a worker client from the original client e.g.

`
future = client.submit(function_returning_a_list, *args)
with client.worker_client(key_timeout=3600) as wc:
    worker_future = wc.map(function_on_one_item, future._result())
`

In a second client session I would then be able to retrieve the worker future via the key from the central datastructure described above. This would have the advantage of giving people an interface more similar to the rest of Dask/Distributed that does not need anything distributed related in the functions being called.

Alternatively perhaps there is something that can be done to extend the dask protocol so it covers situations where an array is generated from a function and then you want to map on that array,

**Comment by mrocklin** [comment]
I think that you could accomplish what you're asking for above with the current implementation and a tiny bit of tornado.

```python
@gen.coroutine
def wait_then_map(future, function):
    results = yield future._result()
    futures = client.map(function, results)
```

**Comment by mrocklin** [comment]
@pitrou do you have any experiences with systems that manage streams of data between multiple actors?  Any general feedback or thoughts on this issue?  My ideas here are not very well informed.

**Comment by mrocklin** [comment]
Also cc @quasiben

**Comment by mrocklin** [comment]
This was implemented in #729
